### ISSUE TITLE
Preact + @tanstack/react-query returns TypeError: Cannot read properties of null (reading 'useEffect')

### ISSUE THREAD

**Issue by xleepy** [vitest-dev/vitest#5004]
### Describe the bug

Hi, 
Trying to use useQuery with preact + vite and cannot run tests because of error 
TypeError: Cannot read properties of null (reading 'useEffect')
Tried to apply different approaches but vitest still fails and looks like cannot apply preact/compat alias but dev/build works as expected. 

### Reproduction

Repository: https://github.com/xleepy/recipes-app
Steps to reproduce:
npm install
npm run generate-api
npm run test

### System Info

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (16) x64 AMD Ryzen 7 5800X3D 8-Core Processor
    Memory: 20.09 GB / 31.91 GB
  Binaries:
    Node: 18.16.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - C:\Program Files\nodejs\yarn.CMD
    npm: 9.5.1 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Chromium (120.0.2210.133)
    Internet Explorer: 11.0.19041.3636
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by hi-ogawa** [comment]
Thanks for the reproduction. The difference of the effect of `resolve.alias` between Vite and Vitest is somewhat expected with a similar caveat as Vite SSR's external https://vitejs.dev/guide/ssr.html#ssr-externals

I haven't tried your repro yet (since it didn't look exactly minimal), but from the quick, maybe some deep dependency could be installing `react` instead of `preact`. Can you maybe try `"overrides": { ... }` https://docs.npmjs.com/cli/v9/configuring-npm/package-json#overrides  to make sure all dependencies sees `preact`?

**Comment by xleepy** [comment]
Will give a try. Thanks!

**Comment by xleepy** [comment]
Update project with overrides but still no luck 
https://github.com/xleepy/recipes-app/commit/8ea1eb16001cdaae3062ec0e2d15e19f7e7ba2f8

**Comment by c-vetter** [comment]
I'm running into this as well in a `react` project, with `vitest@2.0.3` and `vitest@2.0.5`, also `vitest@1.6.0`.

----

Repro based on the `basic` boilerplate:
https://stackblitz.com/edit/vitest-dev-vitest-hps856

Repro based on the `react` boilerplate, with `basic` code copied over:
https://stackblitz.com/edit/vitest-dev-vitest-hi7ytz

----

Any test importing and calling `use*` from the `react` package, directly or via the subject under test, breaks with this error:
> TypeError: Cannot read properties of null (reading 'use*')

And the error location is wrong in each case.

There is one exception in the `react`-based repro: 
```sh
 FAIL  test/basic.test.tsx > useSquared
Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
 ❯ Object.throwInvalidHookError node_modules/react-dom/cjs/react-dom.development.js:15408:9
 ❯ Proxy.useState node_modules/react/cjs/react.development.js:1622:21
 ❯ Module.useSquared src/basic.ts:4:15
      2| 
      3| export function useSquared(n: number) {
      4|   const [x] = useState(n);
       |               ^
      5| 
      6|   return x * x;
 ❯ eval test/basic.test.tsx:15:10
```
Note that this usage follows React's Rules of Hooks, so this might actually point to the cause, although I still cannot seem to determine that myself.

----

Interestingly, this only seems to happen for hooks. `isValidElement` from `react` does not cause this issue.

Additionally note that the `useState inside the `Link` component does not cause the issue either.

----

#4000 describes the same issue and the repro there still shows this: https://stackblitz.com/edit/vitest-dev-vitest-ff6sbs

**Comment by hi-ogawa** [comment]
@c-vetter Your issue is likely different. React simply doesn't allow calling hooks outside of component ("rendering context" or whatever they call it). Here is a quick example and your usages of hooks are essentially same:

```ts
$ node -e 'require("react").useState(0)'
Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://react.dev/link/invalid-hook-call for tips about how to debug and fix this problem.
/home/hiroshi/code/personal/vite-plugins/node_modules/.pnpm/react@19.0.0-rc-df5f2736-20240712/node_modules/react/cjs/react.development.js:1533
      return resolveDispatcher().useState(initialState);
                                ^

TypeError: Cannot read properties of null (reading 'useState')
    at process.env.NODE_ENV.exports.useState (/home/hiroshi/code/personal/vite-plugins/node_modules/.pnpm/react@19.0.0-rc-df5f2736-20240712/node_modules/react/cjs/react.development.js:1533:33)
    at [eval]:1:18
    at runScriptInThisContext (node:internal/vm:209:10)
    at node:internal/process/execution:118:14
    at [eval]-wrapper:6:24
    at runScript (node:internal/process/execution:101:62)
    at evalScript (node:internal/process/execution:133:3)
    at node:internal/main/eval_string:51:3

Node.js v20.14.0
```

It looks like Vitest's error stacktrace is odd, but again that's a different topic from the current issue which is about the intricacy of preact alias and Vite SSR.

**Comment by c-vetter** [comment]
@hi-ogawa You are correct, my issue results from failing to use `renderHook` which I failed to realize at the end of a long work week. A colleague found the issue and everything is working as intended. Thank you for your feedback.

**Comment by Snow31ind** [comment]
@c-vetter How did you and your colleague resolve it?

**Comment by hi-ogawa** [comment]
Package manager level overrides is required since `"@tanstack/react-query` depends on `react` transitively. I tested a following with a minimal repro provided in https://github.com/TanStack/query/issues/8555#issue-2800538950.

https://stackblitz.com/edit/github-ykwyclo2?file=package.json&terminal=test

```js
  "dependencies": {
    "react": "npm:@preact/compat",
    "react-dom": "npm:@preact/compat",
  },
  "overrides": {
    "react": "npm:@preact/compat",
    "react-dom": "npm:@preact/compat"
  },
```

Please provide a reproduction if something is not working.
